crates/dht/src/time.rs

```diff
@@ -0,0 +1,1034 @@
+            next_update_at: Timestamp::from_micros(0),
+        }
+    }
+
+    /// The timestamp at which the last full slice ends.
+    ///
+    /// If there are no full slices, then this is the origin timestamp.
+    fn full_slice_end_timestamp(&self) -> Timestamp {
+        let full_slices_duration = Duration::from_secs(
+            self.full_slices * (1u64 << self.factor) * UNIT_TIME.as_secs(),
+        );
+        self.origin_timestamp + full_slices_duration
+    }
+
+    /// Layout the partial slices for the given time range.
+    ///
+    /// Tries to allocate as many large slices as possible to fill the space.
+    fn layout_partials(
+        &self,
+        current_time: Timestamp,
+        mut start_at: Timestamp,
+    ) -> K2Result<Vec<(Timestamp, u8)>> {
+        let mut recent_time = (current_time - start_at).map_err(|_| {
+            K2Error::other("Failed to calculate recent time for partials, either the clock is is wrong or this is a bug")
+        })?;
+
+        let mut partials = Vec::new();
+
+        // Now we want to partition the remaining time into smaller slices
+        for i in (0..self.factor).rev() {
+            // Starting from the largest slice size, if there's space for two of that slice size
+            // then add two slices of that size, otherwise add one slice of that size
+            let slice_size =
+                Duration::from_secs((1u64 << i) * UNIT_TIME.as_secs());
+            let slice_count =
+                if recent_time > residual_duration_for_factor(i) + slice_size {
+                    2
+                } else if recent_time > slice_size {
+                    1
+                } else {
+                    continue;
+                };
+
+            for _ in 0..slice_count {
+                partials.push((start_at, i));
+                start_at += slice_size;
+                recent_time -= slice_size;
+            }
+        }
+
+        Ok(partials)
+    }
+}
+
+/// Computes what duration is required for a series of time slices.
+///
+/// The duration is computed as the sum of the powers of two from 0 to the factor, multiplied by
+/// the [UNIT_TIME].
+///
+/// Panics if the factor is greater than 53.
+/// Note that the maximum factor changes if the [UNIT_TIME] is changed.
+fn residual_duration_for_factor(factor: u8) -> Duration {
+    if factor >= 54 {
+        panic!("Factor must be less than 54");
+    }
+
+    let mut sum = 1;
+    for i in 1..=factor {
+        sum |= 1u64 << i;
+    }
+
+    Duration::from_secs(sum * UNIT_TIME.as_secs())
+}
+
+/// Combine a series of op hashes into a single hash.
+///
+/// Requires that the op hashes are already ordered.
+/// If the input is empty, then the output is an empty byte array.
+fn combine_op_hashes(hashes: Vec<OpId>) -> bytes::Bytes {
+    hashes
+        .into_iter()
+        .map(|x| x.0 .0)
+        .reduce(|a, b| a.iter().zip(b.iter()).map(|(a, b)| a ^ b).collect())
```

Comment:
nitpick, but I'm not sure functional/reduce is right here... it's going to keep re-allocating memory. It would be more efficient to have a mut [BytesMut](https://docs.rs/bytes/1.8.0/bytes/struct.BytesMut.html) and freeze it once at the end into a Bytes.

Reply:
Yep, perfect!